crates/red_knot/src/symbols.rs

```diff
@@ -579,15 +584,30 @@ struct FlowNodeId;
 enum FlowNode {
     Start,
     Definition(DefinitionFlowNode),
+    Branch(BranchFlowNode),
+    Phi(PhiFlowNode),
 }
 
+/// A Definition node represents a point in control flow where a symbol is defined
 #[derive(Debug)]
 struct DefinitionFlowNode {
     symbol_id: SymbolId,
     definition: Definition,
     predecessor: FlowNodeId,
 }
 
+/// A Branch node represents a branch in control flow
+#[derive(Debug)]
+struct BranchFlowNode {
+    predecessor: FlowNodeId,
+}
+
+/// A Phi node represents a join point where control flow paths come together
+#[derive(Debug)]
+struct PhiFlowNode {
+    predecessors: Vec<FlowNodeId>,
```

Comment:
I'm not sure how the `unreachable` PR solved this but having a `Vec` here is somewhat expensive because we now have to allocate whenever two branches join, and that's very common. 

It might be worth to spend some time exploring how we can improve this by:

* Use an arena allocator for the `Vec`s (least preferred)
* Change the `Phi` node to only ever join 2 branches, and chain them together
* Store the predecessors in a centralized storage and only store the `Range` in the `PhiFlowNode` into that `Vec` (requires that we can write all predecessor entries at once).

Reply:
Switched to exactly two predecessors per Phi node, no more vec. Thanks for the suggestion!